File: backend/app/models/technician_availability.py
from uuid import UUID
from typing import Dict, Any, Optional
from pydantic import BaseModel

from .base import TimeSlotDay


class BaseTechnicianAvailability(BaseModel):
    technician_id: UUID
    timeslot: TimeSlotDay


class TechnicianAvailabilityInDB(BaseTechnicianAvailability):
    id: UUID
    active: bool


class TechnicianAvailabilityCreate(BaseTechnicianAvailability):
    class Config:
        json_schema_extra: Dict[str, Any] = {
            "example": {
                "technician_id": "123e4567-e89b-12d3-a456-426655440000",
                "timeslot": {
                    "day": 1,
                    "start_time": "10:00:00",
                    "end_time": "12:00:00",
                }
            }
        }


class TechnicianAvailabilityUpdate(BaseModel):
    timeslot: TimeSlotDay

    class Config:
        json_schema_extra: Dict[str, Any] = {
            "example": {
                "timeslot": {
                    "day": 2,
                    "start_time": "10:00:00",
                    "end_time": "12:00:00",
                    "active": False
                },
            }
        }


class TechnicianAvailabilityResponse(TechnicianAvailabilityInDB):
    json_schema_extra: Dict[str, Any] = {
        "example": {
            "id": "123e4567-e89b-12d3-a456-426655440000",
            "technician_id": "123e4567-e89b-12d3-a456-426655440000",
            "timeslot": {
                "day": 3,
                "start_time": "10:00:00",
                "end_time": "12:00:00",
            },
            "active": True
        }
    }

File: frontend/src/lib/types/technician_availability.ts
import { z } from "zod";
import { TimeSlotDaySchema } from "./base/timeslot";

const BaseTechnicianAvailabilitySchema = z.object({
    technician_id: z.string().uuid(),
    timeslot: TimeSlotDaySchema
});

export const TechnicianAvailabilityCreateSchema = BaseTechnicianAvailabilitySchema.extend({});

export const TechnicianAvailabilityUpdateSchema = z.object({
    timeslot: TimeSlotDaySchema.optional(),
    active: z.boolean().optional(),
});

export const TechnicianAvailabilityResponseSchema = BaseTechnicianAvailabilitySchema.extend({
    id: z.string().uuid(),
    active: z.boolean()
});

export type TechnicianAvailabilityCreate = z.infer<typeof TechnicianAvailabilityCreateSchema>;
export type TechnicianAvailabilityUpdate = z.infer<typeof TechnicianAvailabilityUpdateSchema>;
export type TechnicianAvailabilityResponse = z.infer<typeof TechnicianAvailabilityResponseSchema>;
